model/src/app_data.rs

```diff
@@ -0,0 +1,132 @@
+//! Contains the app_data file structures to define additional data about tx origin
+
+use crate::h160_hexadecimal::{self};
+use anyhow::{anyhow, Result};
+use cid::multihash::{Code, MultihashDigest};
+use primitive_types::{H160, H256};
+use serde::{Deserialize, Serialize};
+use serde_with::serde_as;
+use std::convert::TryInto;
+
+#[serde_as]
+#[derive(Eq, PartialEq, Clone, Debug, Deserialize, Serialize, Hash, Default)]
+pub struct Referrer {
+    #[serde(with = "h160_hexadecimal")]
+    pub address: H160,
+    pub version: String,
+}
+
+#[serde_as]
+#[derive(Eq, PartialEq, Clone, Debug, Deserialize, Serialize, Hash, Default)]
+pub struct Metadata {
+    pub referrer: Option<Referrer>,
```

Comment:
Is both metadata and referrer really independently optional? It just feels it bloats the code with a lot of unwraps.

Reply:
Yes, this was the spec provided. This gives us the most flexibility in the future. Not sure whether all of that is needed at all, but at least this is what was specified and agreed upon

Reply:
For the DB it will probably make sense to have a table mapping users to users (Referal table). You add a referer there if the user included the referer in his first order.

Anyways, I understand this specific part is not related to the DB, is just a data object modeling a metadata attached to an order so looks good to me.

Reply:
> only the first trade of the users that arrive with a link will get this

Actually, that comes in the order, so there can happen that an order has a referral but a trade is executed for another order without it.

Reply:
> For the DB it will probably make sense to have a table mapping users to users (Referal table). 

I was hoping we can compute most of these queries on the fly and only need to create materialised views when the performance need arises. 

Can a specific user only be referred by one address? What if I click a referral link three months after having first used CowSwap?

Reply:
The idea so far:
1. Referral only applies to new users: someone who traded before is not qualified
2. Only the first referral address used will be considered: subsequent referrals are ignored

So essentially we'll look for referral only on the first trade of an user